server/market/balancer.go

```diff
@@ -77,43 +139,123 @@ func (b *DEXBalancer) CheckBalance(acctAddr string, assetID uint32, qty, lots ui
 	log.Tracef("balance check for %s - %s: new qty = %d, new lots = %d, new redeems = %d",
 		backedAsset.assetInfo.Symbol, acctAddr, qty, lots, redeems)
 
+	// Make sure we can get the primary balance first.
 	bal, err := backedAsset.balancer.AccountBalance(acctAddr)
 	if err != nil {
 		log.Error("(*DEXBalancer).CheckBalance: error getting account balance for %q: %v", acctAddr, err)
 		return false
 	}
+	if bal == 0 {
+		log.Tracef("(*DEXBalancer).CheckBalance(%q, %d, %d, %d, %d) false for zero balance",
+			acctAddr, assetID, qty, lots, redeems)
+		return false
+	}
+
+	// Set up fee tracking for tokens.
+	var feeID uint32 // Asset ID of the base chain asset.
+	// feeQty will track the fee asset order-locked quantity, but is only used
+	// for tokens. When not a token (when assetID = base chain), the
+	// order-locked quantity is summed in reqFunds as for any other asset. We'll
+	// fetch feeBal to catch any zero bals or other errors early before
+	// iterating markets and querying the Swapper.
+	var feeQty, feeBal uint64 // Current balance.
+	feeBalancer := backedAsset.feeBalancer
+	isToken := feeBalancer != nil
+	if isToken {
+		feeID = feeBalancer.assetInfo.ID
+		feeBal, err = feeBalancer.balancer.AccountBalance(acctAddr)
+		if err != nil {
+			log.Error("(*DEXBalancer).CheckBalance: error getting fee asset balance for %q: %v", acctAddr, err)
+			return false
+		}
+		if feeBal == 0 {
+			log.Tracef("(*DEXBalancer).CheckBalance(%q, %d, %d, %d, %d) false for zero fee asset (%s) balance",
+				acctAddr, assetID, qty, lots, redeems, feeBalancer.assetInfo.Symbol)
+			return false
+		}
+	}
+
+	var swapFees, redeemFees uint64
```

Comment:
No need to track them separately since they are always summed in the end.

Reply:
`redeemFees` is logged separately. Do you mind if I leave it? I think it helps with clarity.

Reply:
Ah yeah np.